walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,107 @@
+    Iterator<TimeRange> iterator = unavailableTime.iterator();
+    TimeRange currentEvent = (iterator.hasNext()) ?  iterator.next() : null;
+    if (currentEvent != null) {
+      if (TimeRange.START_OF_DAY < currentEvent.start()
+          && currentEvent.start() - TimeRange.START_OF_DAY >= request.getDuration()) {
+        validTimeRanges.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, currentEvent.start(), false));
+      }
+
+      while (iterator.hasNext()) {
+        TimeRange nextEvent = iterator.next();
+
+        if (nextEvent.start() - currentEvent.end() >= request.getDuration()) {
+          validTimeRanges.add(TimeRange.fromStartEnd(currentEvent.end(), nextEvent.start(), false));
+        }
+        currentEvent = nextEvent;
+      }
+
+      if (currentEvent.end() < TimeRange.END_OF_DAY
+          && TimeRange.END_OF_DAY - currentEvent.end() + 1 >= request.getDuration()) {
+        validTimeRanges.add(TimeRange.fromStartEnd(currentEvent.end(), TimeRange.END_OF_DAY, true));
+      }
+    } else {  // There are no times taken in the day.
+      validTimeRanges.add(TimeRange.WHOLE_DAY);
+    }
+    return validTimeRanges;
   }
+
+  /**
+   * @return a Set containing all the unavailable time in the day.
+   */
+  private Set<TimeRange> determineUnavailableTime(Collection<Event> events, MeetingRequest request) {
+    Collection<String> attendees = request.getAttendees();
+    TreeSet<TimeRange> timeTaken = new TreeSet<>((range1, range2) -> {
+      if (range1.start() < range2.start()) {
+        return -1;
+      } else if (range1.start() > range2.start()) {
+        return 1;
+      } else {
+        return 0;
+      }
+    });
+
+    for (String attendee : attendees) {
+      for (Event event : events) {
+        if (event.getAttendees().contains(attendee)) {
+          timeTaken.add(event.getWhen());
+        }
+      }
+    }
+    return combineRanges(timeTaken);
+  }
+
+  /**
+   * Combines all ranges that overlap or are back-to-back into one range.
+   */
+  private TreeSet<TimeRange> combineRanges(TreeSet<TimeRange> timeRanges) {
+    Iterator<TimeRange> iterator = timeRanges.iterator();
+    TimeRange firstRange = (iterator.hasNext()) ? iterator.next() : null;
+
+    while (firstRange != null && iterator.hasNext()) {
+      TimeRange toCombine = iterator.next();
+      if (firstRange.overlaps(toCombine) || firstRange.end() == toCombine.start()) {
+        timeRanges.remove(firstRange);
+        timeRanges.remove(toCombine);
+        timeRanges.add(TimeRange.fromStartEnd(
+            Math.min(firstRange.start(), toCombine.start()),
+            Math.max(firstRange.end(), toCombine.end()),
+            false));
+
+        // Reset iterator to contain the newly combined piece.
+        iterator = timeRanges.iterator();
+        firstRange = iterator.next();
+      } else {
+        firstRange = toCombine;  // Store current range for comparing with the next.
```

Comment:
[Nit] I might call the 'currentRange' and 'nextRange' as that makes it more clear what their purpose is after the first iterator of the loop.